qrenderdoc/Windows/Dialogs/PerformanceCounterSelection.cpp

```diff
@@ -0,0 +1,290 @@
+    if(f.open(QIODevice::ReadOnly | QIODevice::Text))
+    {
+      LoadFromJSON(doc, f, JSON_ID, JSON_VER);
+
+      std::set<Uuid> selectedCounters;
+
+      QVariantList counters = doc[QLatin1String{"counters"}].toList();
+
+      for(const auto &counter : counters)
+      {
+        QVariantList bytes = counter.toList();
+        Uuid uuid;
+
+        /// TODO assert counter.size () == 4
+
+        for(int i = 0; i < 4; ++i)
+        {
+          uuid.bytes[i] = bytes[i].toUInt();
+        }
+
+        selectedCounters.insert(uuid);
+      }
+
+      // We we walk over the complete tree, and toggle everything so it
+      // matches the settings
+      QTreeWidgetItemIterator it(ui->counterTree);
+      while(*it)
+      {
+        const auto id = (*it)->data(0, Qt::UserRole + 2);
+        if(id.isValid())
+        {
+          const GPUCounter counter = (GPUCounter)id.toUInt();
+
+          if(!m_CounterToUuid.contains(counter))
+            continue;
+
+          (*it)->setCheckState(
+              0, (selectedCounters.find(m_CounterToUuid[counter]) != selectedCounters.end())
+                     ? Qt::Checked
+                     : Qt::Unchecked);
+        }
+
+        // The loop above will uncheck all unknown counters, and not crash if some counter is no
+        // longer present, or unknown
+
+        ++it;
+      }
+    }
+    else
+    {
+      RDDialog::critical(this, tr("Error loading config"),
+                         tr("Couldn't open path %1 for reading.").arg(filename));
+    }
+
+  });
+
+  connect(ui->counterTree, &QTreeWidget::itemChanged, [&](QTreeWidgetItem *item, int) -> void {
+    const auto d = item->data(0, Qt::UserRole + 2);
+
+    if(d.isValid())
+    {
+      if(item->checkState(0) == Qt::Checked)
+      {
+        // Add
+        auto listItem = new QListWidgetItem(ui->enabledCounters);
+        listItem->setText(item->text(0));
+        m_SelectedCounters.insert((GPUCounter)d.toUInt(), listItem);
+      }
+      else
+      {
+        // Remove
+        auto listItem = m_SelectedCounters.take((GPUCounter)d.toUInt());
+        delete listItem;
+      }
+    }
+  });
+
+  connect(ui->sampleCounters, &QPushButton::pressed, [&]() -> void { this->accept(); });
+
+  ui->counterTree->setMouseTracking(true);
+
+  auto showCounters = [&](const QVector<CounterDescription> &counters) -> void {
+    ui->counterTree->clear();
+    ui->enabledCounters->clear();
+
+    QTreeWidgetItem *currentRoot = NULL;
+    CounterFamily currentFamily = CounterFamily::Unknown;
+
+    std::unordered_map<std::string, QTreeWidgetItem *> categories;
+
+    for(const auto desc : counters)
+    {
+      m_CounterToUuid[desc.counterID] = desc.uuid;
+      m_UuidToCounter[desc.uuid] = desc.counterID;
+
+      const CounterFamily family = GetCounterFamily(desc.counterID);
+      if(family != currentFamily)
+      {
+        currentRoot = new QTreeWidgetItem(ui->counterTree);
+        currentRoot->setText(0, QLatin1String{ToString(family)});
```

Comment:
If it's to be immediately converted to `QString` it'd be better to have `ToString` return a `QString` and then the conversion can be at compile-time with `return lit("AMD")` etc.